src/fixing/mod.rs

```diff
@@ -0,0 +1,350 @@
+use std::{
+    cmp::Ordering,
+    collections::HashMap,
+    ops::Deref,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use dashmap::DashMap;
+use squalid::OptionExt;
+use tracing::{debug, debug_span, instrument};
+use tree_sitter_grep::{
+    get_parser,
+    tree_sitter::{InputEdit, Point, Range, Tree},
+    Parseable, RopeOrSlice, SupportedLanguageLanguage,
+};
+
+use crate::{
+    aggregated_queries::AggregatedQueries, rule::InstantiatedRule, run_per_file, Config,
+    FileRunContext, FromFileRunContextInstanceProviderFactory, MutRopeOrSlice, RuleMeta, RuleName,
+    RunKind, ViolationWithContext,
+};
+
+mod accumulated_edits;
+mod fixer;
+
+pub use accumulated_edits::{get_newline_offsets_rope_or_slice, AccumulatedEdits};
+pub use fixer::{Fixer, PendingFix};
+
+const MAX_FIX_ITERATIONS: usize = 10;
+
+#[allow(clippy::too_many_arguments)]
+#[instrument(level = "debug", skip_all, fields(?path))]
+pub fn run_fixing_loop<'a>(
+    violations: &mut Vec<ViolationWithContext>,
+    file_contents: impl Into<MutRopeOrSlice<'a>>,
+    mut pending_fixes: HashMap<RuleName, (Vec<PendingFix>, Arc<RuleMeta>)>,
+    aggregated_queries: &AggregatedQueries,
+    path: &Path,
+    config: &Config,
+    language: SupportedLanguageLanguage,
+    instantiated_rules: &[InstantiatedRule],
+    tree: Tree,
+    from_file_run_context_instance_provider_factory: &dyn FromFileRunContextInstanceProviderFactory,
+    initial_run_kind: RunKind,
+) -> AccumulatedEdits {
+    let mut file_contents = file_contents.into();
+    let mut old_tree = tree;
+    let mut accumulated_edits = AccumulatedEdits::new(
+        get_newline_offsets_rope_or_slice(RopeOrSlice::from(&file_contents)).collect(),
+    );
+    for _ in 0..MAX_FIX_ITERATIONS {
+        let _span = debug_span!("single fixing loop pass").entered();
+
+        let input_edits_and_replacements = apply_fixes(&mut file_contents, pending_fixes);
+        for (input_edit, _) in &input_edits_and_replacements {
+            old_tree.edit(input_edit);
+        }
+        accumulated_edits.add_round_of_edits(&input_edits_and_replacements);
+
+        if config.single_fixing_pass {
+            if !config.report_fixed_violations {
+                *violations = violations
+                    .iter()
+                    .filter(|violation| !violation.had_fixes)
+                    .cloned()
+                    .collect();
+            }
+            return accumulated_edits;
```

Comment:
Ok so I think the idea is that all of the edits from the fixing loop are getting tracked (despite the fact that in non-initial fixing loop passes the location data might have "shifted" relative to the original location data of the corresponding place in the file) so that that can get passed to the LSP server (for optimization-y stuff)?